pkg/dbs/studydb/db_study-surveys.go

```diff
@@ -13,56 +15,132 @@ func (dbService *StudyDBService) SaveSurvey(instanceID string, studyKey string,
 	ctx, cancel := dbService.getContext()
 	defer cancel()
 
-	filter := bson.M{"current.surveyDefinition.key": survey.Current.SurveyDefinition.Key}
+	res, err := dbService.collectionRefStudySurveys(instanceID, studyKey).InsertOne(ctx, survey)
+	survey.ID = res.InsertedID.(primitive.ObjectID)
+	return survey, err
+}
+
+func (dbService *StudyDBService) UnpublishSurvey(instanceID string, studyKey string, surveyKey string) error {
+	ctx, cancel := dbService.getContext()
+	defer cancel()
 
-	upsert := true
-	rd := options.After
-	options := options.FindOneAndReplaceOptions{
-		Upsert:         &upsert,
-		ReturnDocument: &rd,
+	filter := bson.M{
+		"surveyDefinition.key": surveyKey,
+		"unpublished":          0,
 	}
-	elem := types.Survey{}
-	err := dbService.collectionRefStudySurveys(instanceID, studyKey).FindOneAndReplace(
-		ctx, filter, survey, &options,
-	).Decode(&elem)
-	return elem, err
+	update := bson.M{"$set": bson.M{"unpublished": time.Now().Unix()}}
+	_, err := dbService.collectionRefStudySurveys(instanceID, studyKey).UpdateMany(ctx, filter, update)
+	return err
 }
 
-func (dbService *StudyDBService) RemoveSurveyFromStudy(instanceID string, studyKey string, surveyKey string) error {
+func (dbService *StudyDBService) GetSurveyKeysInStudy(instanceID string, studyKey string, includeUnpublished bool) (surveyKeys []string, err error) {
 	ctx, cancel := dbService.getContext()
 	defer cancel()
 
-	filter := bson.M{"current.surveyDefinition.key": surveyKey}
-	res, err := dbService.collectionRefStudySurveys(instanceID, studyKey).DeleteOne(ctx, filter)
-	if res.DeletedCount < 1 {
-		err = errors.New("not found")
+	filter := bson.M{}
+	if !includeUnpublished {
+		filter["unpublished"] = 0
 	}
-	return err
+	res, err := dbService.collectionRefStudySurveys(instanceID, studyKey).Distinct(ctx, "surveyDefinition.key", filter)
+	if err != nil {
+		return surveyKeys, err
+	}
+	surveyKeys = make([]string, len(res))
+	for i, r := range res {
+		surveyKeys[i] = r.(string)
+	}
+	return surveyKeys, err
+
 }
 
-func (dbService *StudyDBService) FindSurveyDef(instanceID string, studyKey string, surveyKey string) (types.Survey, error) {
+func (dbService *StudyDBService) FindAllCurrentSurveyDefsForStudy(instanceID string, studyKey string, includeUnpublished bool, onlyInfos bool) (surveys []*types.Survey, err error) {
+	surveyKeys, err := dbService.GetSurveyKeysInStudy(instanceID, studyKey, includeUnpublished)
+	if err != nil {
+		return surveys, err
+	}
+	for _, key := range surveyKeys {
+		survey, err := dbService.FindCurentSurveyDef(instanceID, studyKey, key, onlyInfos)
+		if err != nil {
+			logger.Error.Println(err)
+			continue
```

Comment:
If one of these calls fail, is it really safe to continue and return an incomplete list of surveys?

Reply:
Ok, I think the current way the method "FindAllCurrentSurveyDefsForStudy" is implemented is flawed, because, we can get also survey keys that are unpublished, but "FindCurrentSurveyDef" would return with error. 

Need to rework this.

Reply:
So, removed the flag to include unpublished, since that would not work. This way, an error is an unexpected event, most probably some issue with the database connection / or data base itself, and is a temporary issue. So iterrupting and returning this event is probably a good idea.